src/app/widgets/widget-racer-line/widget-racer-line.component.ts

```diff
@@ -0,0 +1,405 @@
+        break;
+      case 'green':
+        this.labelColor = this.theme().greenDim;
+        this.valueColor = this.theme().green;
+        break;
+      case 'pink':
+        this.labelColor = this.theme().pinkDim;
+        this.valueColor = this.theme().pink;
+        break;
+      case 'orange':
+        this.labelColor = this.theme().orangeDim;
+        this.valueColor = this.theme().orange;
+        break;
+      case 'purple':
+        this.labelColor = this.theme().purpleDim;
+        this.valueColor = this.theme().purple;
+        break;
+      case 'grey':
+        this.labelColor = this.theme().greyDim;
+        this.valueColor = this.theme().grey;
+        break;
+      case 'yellow':
+        this.labelColor = this.theme().yellowDim;
+        this.valueColor = this.theme().yellow;
+        break;
+      default:
+        this.labelColor = this.theme().contrastDim;
+        this.valueColor = this.theme().contrast;
+        break;
+    }
+    this.dtsColor = this.valueColor;
+  }
+
+  ngOnDestroy() {
+    console.log('ngOnDestroy!');
+    this.isDestroyed = true;
+    this.destroyDataStreams();
+    this.canvasService.clearCanvas(this.dToLineContext, this.dToLineElement.width, this.dToLineElement.height);
+  }
+
+  private updateCanvas(): void {
+    this.drawDToLine();
+    this.drawLenBias();
+    this.drawUnit();
+  }
+
+  private drawDToLine(): void {
+    if (this.dToLineCanvas) {
+      this.canvasService.clearCanvas(this.dToLineContext, this.dToLineElement.width, this.dToLineElement.height);
+      const valueText = this.getValueText();
+      this.canvasService.drawText(
+        this.dToLineContext,
+        valueText,
+        Math.floor(this.dToLineElement.width / 2),
+        Math.floor((this.dToLineElement.height / 2) * 1.15),
+        this.maxValueTextWidth,
+        this.maxValueTextHeight,
+        'bold',
+        this.dtsColor
+      );
+    }
+  }
+
+  private getValueText(): string {
+    if (this.dtsValue === null) {
+      return '--';
+    }
+    return this.dtsValue.toFixed(this.widgetProperties.config.numDecimal);
+  }
+
+  private drawUnit(): void {
+    const unit = this.widgetProperties.config.paths['dtsPath'].convertUnitTo;
+    this.canvasService.drawText(
+      this.dToLineContext,
+      unit,
+      Math.floor(this.dToLineElement.width - 10 * this.canvasService.scaleFactor),
+      Math.floor(this.dToLineElement.height - 10 * this.canvasService.scaleFactor),
+      Math.floor(this.dToLineElement.width * 0.25),
+      Math.floor(this.dToLineElement.height * 0.15),
+      'bold',
+      this.dtsColor,
+      'end',
+      'alphabetic'
+    );
+  }
+
+  private drawLenBias(): void {
+    if (this.widgetProperties.config.paths['lineLengthPath'].path !== '') {
+      const unit = this.widgetProperties.config.paths['lineLengthPath'].convertUnitTo;
+      this.lineLengthValue = `―${this.applyDecorations(this.lengthValue.toFixed(this.widgetProperties.config.numDecimal))}${unit}―`;
```

Comment:
This line generates: ERROR TypeError: Cannot read properties of null (reading 'toFixed'). It's because drawLenBias() get's called before this.lengthValue receives data.

drawLenBias() can be called in these 3 places and the first to get data calls updateCanvas() so you will have null variables. You can either initialize with a value, or split the draw method per the value to update:
- this.observeDataStream('dtsPath', newValue => {
- this.observeDataStream('lineLengthPath', newValue => {
- this.observeDataStream('lineBiasPath', newValue => {